### PR TITLE
XFA - Add a lexer/parser for FormCalc language

### PR DESCRIPTION
**Pull request by calixteman**
- the language specifications are: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=1049
 - it can be used to:
   * as a scripting language for calculation, validations, ...
   * in SOM expressions to select nodes: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=101

### CODE REVIEW

**Comment by calixteman** [comment]
For the scripting stuff, an idea is to generate some javascript from the generated AST and then execute it in the js sandbox.
It's the way chose by pdfium:
https://pdfium.googlesource.com/pdfium/+/refs/heads/master/xfa/fxfa/fm2js/cxfa_fmexpression.cpp

**Comment by benbro** [comment]
Will it be possible to disable scripting?

**Comment by timvandermeij** [comment]
It's a preference now and it's disabled by default. There are no plans to remove this preference because scripting is a feature that we can imagine that people want to disable.

This question is unrelated to this PR though, so let's leave it at this.

File: src/core/xfa/formcalc_lexer.js [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    if (0x30 /* = 0 */ <= char && char <= 0x39 /* = 9 */) {
+      this.pos++;
+      const match = this.data.substring(this.pos).match(dotNumberPattern);
+      if (!match) {
+        return new Token(TOKEN.number, (char - 0x30) /* = 0 */ / 10);
+      }
+      const end = this.pos + match[0].length;
+      const number = parseFloat(this.data.substring(this.pos - 2, end));
```
Comment (brendandahl) [inline src/core/xfa/formcalc_lexer.js:295]: What should happen in this language if we parse a really large number that would be infinity in JS? e.g. "9e99999"

**Review by brendandahl — CHANGES_REQUESTED** [review]

File: src/core/xfa/formcalc_lexer.js [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+      return new Token(alt1);
+    }
+    return new Token(alt2);
+  }
+
+  getLower() {
+    const char = this.data.charCodeAt(this.pos);
+    if (char === 0x3d /* = = */) {
+      this.pos++;
+      return new Token(TOKEN.le);
```
[the remark this answers is not in the document]
Comment (calixteman) [reply src/core/xfa/formcalc_lexer.js:252]: good idea

File: src/core/xfa/formcalc_lexer.js (hunk 1 above)
Comment (calixteman) [reply src/core/xfa/formcalc_lexer.js:295]: Numbers are following IEEE754 format: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=1056
like in JS: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Numbers_and_dates
So using `parseFloat` should give the correct number.

**Review by brendandahl — APPROVED** [review]
r+ with tests passing

**Pull request merged**